libsql-sqlite3/src/vectorIndex.c

```diff
@@ -0,0 +1,462 @@
+/*
+** 2024-03-18
+**
+** Copyright 2024 the libSQL authors
+**
+** Permission is hereby granted, free of charge, to any person obtaining a copy of
+** this software and associated documentation files (the "Software"), to deal in
+** the Software without restriction, including without limitation the rights to
+** use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+** the Software, and to permit persons to whom the Software is furnished to do so,
+** subject to the following conditions:
+**
+** The above copyright notice and this permission notice shall be included in all
+** copies or substantial portions of the Software.
+**
+** THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+** IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+** FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+** COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+** IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+** CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+**
+******************************************************************************
+**
+** libSQL vector search.
+*/
+#ifndef SQLITE_OMIT_VECTOR
+#include "sqlite3.h"
+#include "vdbeInt.h"
+#include "sqliteInt.h"
+#include "vectorIndexInt.h"
+
+/**************************************************************************
+** VectorIdxParams utilities
+****************************************************************************/
+
+void vectorIdxParamsInit(VectorIdxParams *pParams, u8 *pBinBuf, int nBinSize) {
+  assert( nBinSize <= VECTOR_INDEX_PARAMS_BUF_SIZE );
+
+  pParams->nBinSize = nBinSize;
+  if( pBinBuf != NULL ){
+    memcpy(pParams->pBinBuf, pBinBuf, nBinSize);
+  }
+}
+
+u64 vectorIdxParamsGetU64(const VectorIdxParams *pParams, char tag) {
+  int i, offset;
+  u64 value = 0;
+  for (i = 0; i + 9 <= pParams->nBinSize; i += 9){
+    if( pParams->pBinBuf[i] != tag ){
+      continue;
+    }
+    // choose latest value from the VectorIdxParams bin
+    value = 0;
+    for(offset = 0; offset < 8; offset++){
+      value |= ((u64)(pParams->pBinBuf[i + 1 + offset]) << (u64)(8 * offset));
+    }
+  }
+  return value;
```

Comment:
Would it make sense to go backwards and stop at first element that has the right tag?

Reply:
Thats actually open question. Initially I thought that current logic will allow us easily implement "overrides" - so it has bit more flexibility in a sense.

But now I'm actually not sure if it's actually useful feature of the format...

Reply:
So going backwards and stoping at the first element with the right tag would be functionally the same as the code in the PR it would just maybe be simpler/more performant as we would compose value only once.

Reply:
I decided to leave current implementation as is.

Actually, since I want this parsing method to be safe, it became bit tricker as you need to calculate starting offset from the end (I don't want to rely on fact that payloads always will be multiple of 9; maybe in future we will add more compact parameters representation which will be controlled by tag bits).

Reply:
Makes senes.